x-pack/agent/pkg/agent/application/filters/constraints_filter.go

```diff
@@ -0,0 +1,204 @@
+
+package filters
+
+import (
+	"fmt"
+	"runtime"
+
+	"github.com/Masterminds/semver"
+
+	"github.com/elastic/beats/v7/x-pack/agent/pkg/agent/application/info"
+	"github.com/elastic/beats/v7/x-pack/agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/agent/pkg/agent/transpiler"
+	"github.com/elastic/beats/v7/x-pack/agent/pkg/boolexp"
+	"github.com/elastic/beats/v7/x-pack/agent/pkg/release"
+	"github.com/elastic/go-sysinfo"
+)
+
+const (
+	datasourcesKey          = "datasources"
+	constraintsKey          = "constraints"
+	validateVersionFuncName = "validate_version"
+)
+
+const (
+	agentIDKey      = "agent.id"
+	agentVersionKey = "agent.version"
+	hostArchKey     = "host.architecture"
+	osFamilyKey     = "os.family"
+	osKernelKey     = "os.kernel"
+	osPlatformKey   = "os.platform"
+	osVersionKey    = "os.version"
+)
+
+// ConstraintFilter filters ast based on included constraints.
+func ConstraintFilter(ast *transpiler.AST) error {
+	// get datasources
+	dsNode, found := transpiler.Lookup(ast, datasourcesKey)
+	if !found {
+		return nil
+	}
+
+	dsListNode, ok := dsNode.Value().(*transpiler.List)
+	if !ok {
+		return nil
+	}
+
+	dsList, ok := dsListNode.Value().([]transpiler.Node)
+	if !ok {
+		return nil
+	}
+
+	// for each datasource
+	i := 0
+	originalLen := len(dsList)
+	for i < len(dsList) {
+		constraintMatch, err := evaluateConstraints(dsList[i])
+		if err != nil {
+			return err
+		}
+
+		if constraintMatch {
+			i++
+			continue
+		}
+		dsList = append(dsList[:i], dsList[i+1:]...)
+	}
+
+	if len(dsList) == originalLen {
+		return nil
+	}
+
+	// Replace datasources with limited set
+	if err := transpiler.RemoveKey(datasourcesKey).Apply(ast); err != nil {
+		return err
+	}
+
+	newList := transpiler.NewList(dsList)
+	return transpiler.Insert(ast, newList, datasourcesKey)
+}
+
+func evaluateConstraints(datasourceNode transpiler.Node) (bool, error) {
+	constraintsNode, found := datasourceNode.Find(constraintsKey)
+	if !found {
+		return true, nil
+	}
+
+	constraintsListNode, ok := constraintsNode.Value().(*transpiler.List)
+	if !ok {
+		return false, errors.New("constraints not a list", errors.TypeConfig)
+	}
+
+	constraintsList, ok := constraintsListNode.Value().([]transpiler.Node)
+	if !ok {
+		return false, errors.New("constraints not a list", errors.TypeConfig)
+	}
+
+	for _, c := range constraintsList {
+		strval, ok := c.(*transpiler.StrVal)
+		if !ok {
+			return false, errors.New("constraints is not a string")
+		}
+
+		if isOK, err := evaluateConstraint(strval.String()); !isOK || err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+func evaluateConstraint(constraint string) (bool, error) {
+	regs := boolexp.NewMethodsReg()
+	if err := regs.Register(validateVersionFuncName, regValidateVersion); err != nil {
+		return false, err
+	}
+
+	store, err := newVarStore()
+	if err != nil {
+		return false, err
+	}
+
+	if err := initVarStore(store); err != nil {
```

Comment:
Looking a the fields used I think this could be defined once and reused?

The following requires a reboot to change:
```
	// 	Agent
	store.vars[agentIDKey] = agentInfo.AgentID()
	store.vars[agentVersionKey] = release.Version()

	// Host
	store.vars[hostArchKey] = info.Architecture

	// Operating system
	store.vars[osFamilyKey] = runtime.GOOS
	store.vars[osKernelKey] = info.KernelVersion
	store.vars[osPlatformKey] = info.OS.Family
	store.vars[osVersionKey] = info.OS.Version